persist/fs/persist_manager.go

```diff
@@ -230,15 +231,28 @@ func (pm *persistManager) Prepare(
 		return prepared, errPersistManagerCannotPrepareNotFlushing
 	}
 
-	// NB(xichen): if the checkpoint file for blockStart already exists, bail.
-	// This allows us to retry failed flushing attempts because they wouldn't
-	// have created the checkpoint file.
-	if FilesetExistsAt(pm.filePathPrefix, nsID, shard, blockStart) {
+	// If the checkpoint file aleady exists, bail. This allows us to retry failed attempts because
+	// they wouldn't have created the checkpoint file. This can happen in a variety of situations
+	// for flushes, but is unlikely with snapshots because every snapshot has a unique timestamp
+	// that is not block-aligned.
+	exists, err := pm.filesetExistsAt(opts)
+	if err != nil {
+		return prepared, err
+	}
+	if exists {
 		return prepared, nil
 	}
 
 	blockSize := nsMetadata.Options().RetentionOptions().BlockSize()
-	if err := pm.writer.Open(nsID, blockSize, shard, blockStart); err != nil {
+	writerOpts := WriterOpenOptions{
+		Namespace:    nsID,
+		BlockSize:    blockSize,
+		Shard:        shard,
+		BlockStart:   blockStart,
+		SnapshotTime: snapshotTime,
+		IsSnapshot:   opts.IsSnapshot,
```

Comment:
For this similarly can we make this an enum type instead of a bool?

Reply:
done